### DISCUSSION TITLE
Question about syntax in handlebars

### DISCUSSION THREAD

**Discussion by Kultursylt** [Budibase/budibase#6727] [Help] [answered]
Hi BudiBase

I am a new and rather inexperienced user but I like to be able to build some simpler apps. I am very interested in making conclusions from my data supporting decisions made in real time in the field. Thus calculating on the input becomes crucial (normally I would do this in excel - but you open the possibility to do it in real time)

As I understand it there is no built in support to calculate means in the SQlite nomenclatura. I found an example of code used to do that, but I can not get it to work in your handlebar interface. How does your version of SQlite differ from this?

For example I understand that it has to be started and ended with {{ }}  but as I try it I understand that there are more differences. For example SELECT is not a command. Is there a documentation of this?

<--- I found an example of code made for SQLite that I'd like to use--->
SELECT x
FROM MyTable
ORDER BY x
LIMIT 1
OFFSET (SELECT COUNT(*)
        FROM MyTable) / 2

**Comment by R2bEEaton** [comment, marked as answer]
Hi! I'm a new user as well. Handlebar bindings have nothing to do with SQLite. Handlebars is a templating language that Budibase can use all over the place (including table formulas, SQL queries, component properties, etc.). It uses JavaScript internally, but its syntax is completely different. Handlebars is a framework, you could say, that Budibase uses. An example of Handlebar syntax you find online may not work in Budibase because Budibase uses its own built-in Handlebar helpers (functions) and you can't add custom ones yet.

Guide for Bindings: https://docs.budibase.com/docs/bindings
List of Budibase's specific Handlebar helpers: https://github.com/Budibase/handlebars-helpers
General Handlebar information / examples: https://handlebarsjs.com/guide/ (might not all be applicable to Budibase)

Below is an example of getting the mean of a value from a Budibase data provider with Handlebars (with a little explanation). This should hopefully get you started. This assumes a configuration where you have a "New Data Provider" sourced by some sort of table that includes a numerical "price" column. Inside the New Data Provider element, there is a text component with the text property set to the following:

`{{ avg ( pluck New Data Provider.Rows "price" ) }}`

Just like in arithmetic, let's look inside the parentheses first.
In this case, "pluck" will loop over an array of JSON objects (New Data Provider.Rows) and create an array containing all the "price" key values.
Next, this array is averaged using the "avg" function.
The handlebar expression is surrounded by {{ }} (like handlebar mustaches) so Budibase knows it's a handlebar.

When in doubt, use the bindings search when making handlebars in Budibase; it includes a large amount of functions that can be helpful for transforming your data, even some more complicated things like if...then statements. Also, feel free to just display whatever you're working on. Confused about the pluck syntax? Enter `{{ New Data Provider.Rows }}` and look at the raw JSON. Why isn't avg working? Have a look at the array of values `{{ pluck New Data Provider.Rows "price" }}` and see if there are some non-numerical values in there or something. Another useful thing to know is that array counting begins with 0 as in most programming languages: `{{ New Data Provider.Rows.0 }}` will display the JSON for the first row in the data provider.

Handlebars _are_ a little confusing at first, but I was able to pick it up after a week and they are really helpful for making "code" that is easy to read and for others to come back and maintain. Finally, if there's anything you are struggling to do in Handlebars, don't just give up trying to do it. Anywhere you can use Handlebars, you can also use JavaScript. It might be a little messier and annoying to work with bindings in JS, but with these two options anything is possible!
